drivers/usb/uhc/uhc_mcux_common.c

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2024 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <soc.h>
+#include <string.h>
+#include <stdio.h>
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/init.h>
+#include <zephyr/sys/byteorder.h>
+#include <zephyr/drivers/usb/uhc.h>
+
+#include "uhc_common.h"
+#include "usb.h"
+#include "usb_host_config.h"
+#include "usb_host_mcux_drv_port.h"
+#include "uhc_mcux_common.h"
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uhc_mcux, CONFIG_UHC_DRIVER_LOG_LEVEL);
+
+#define PRV_DATA_HANDLE(_handle) CONTAINER_OF(_handle, struct uhc_mcux_data, mcux_host)
+
+int uhc_mcux_lock(const struct device *dev)
+{
+	struct uhc_data *data = dev->data;
+
+	return k_mutex_lock(&data->mutex, K_FOREVER);
+}
+
+int uhc_mcux_unlock(const struct device *dev)
+{
+	struct uhc_data *data = dev->data;
+
+	return k_mutex_unlock(&data->mutex);
+}
+
+int uhc_mcux_control(const struct device *dev, uint32_t control, void *param)
+{
+	struct uhc_mcux_data *priv = uhc_get_private(dev);
+	usb_status_t status;
+
+	status = priv->mcux_if->controllerIoctl(priv->mcux_host.controllerHandle, control, param);
+	if (status != kStatus_USB_Success) {
+		return -EIO;
+	}
+	return 0;
+}
+
+int uhc_mcux_bus_control(const struct device *dev, usb_host_bus_control_t type)
+{
+	return uhc_mcux_control(dev, kUSB_HostBusControl, &type);
+}
+
+/* Controller driver calls this function when device attach. */
+usb_status_t USB_HostAttachDevice(usb_host_handle hostHandle, uint8_t speed, uint8_t hubNumber,
+				  uint8_t portNumber, uint8_t level,
+				  usb_device_handle *deviceHandle)
+{
+	enum uhc_event_type type;
+	struct uhc_mcux_data *priv;
+
+	if (speed == USB_SPEED_HIGH) {
+		type = UHC_EVT_DEV_CONNECTED_HS;
+	} else if (speed == USB_SPEED_FULL) {
+		type = UHC_EVT_DEV_CONNECTED_FS;
+	} else {
+		type = UHC_EVT_DEV_CONNECTED_LS;
+	}
+
+	priv = (struct uhc_mcux_data *)(PRV_DATA_HANDLE(hostHandle));
+	priv->top_device_speed = speed;
+	uhc_submit_event(priv->dev, type, 0);
+	return kStatus_USB_Success;
+}
+
+/* Controller driver calls this function when device detaches. */
+usb_status_t USB_HostDetachDevice(usb_host_handle hostHandle, uint8_t hubNumber, uint8_t portNumber)
+{
+	struct uhc_mcux_data *priv;
+
+	priv = (struct uhc_mcux_data *)(PRV_DATA_HANDLE(hostHandle));
+	uhc_submit_event(priv->dev, UHC_EVT_DEV_REMOVED, 0);
+	uhc_mcux_bus_control(priv->dev, kUSB_HostBusEnableAttach);
+	return kStatus_USB_Success;
+}
+
+/* MCUX Controller driver calls this function to get the device information. */
+usb_status_t USB_HostHelperGetPeripheralInformation(usb_device_handle deviceHandle,
+						    uint32_t infoCode, uint32_t *infoValue)
+{
+	const struct device *dev = usbh_udev_get_dev(deviceHandle);
```

Comment:
Driver must not call anything from the higher layer. Each uhc_transfer has a device address, you have to use it somehow.

Reply:
The purpose of I add `usbh_udev_get_dev` and `usbh_udev_get_address` is to get device's information by controller driver. I think the device's related hub information (hub address, port num, first hs hub address, first hs hub port num) and speed are needed by the controller driver to do the transfer. Will you plan to put all the information to every `uhc_transfer`?

Reply:
Regardless of whether this information is needed or not, you cannot include this header or use any higher-level API in the driver.

Reply:
Even the linux EHCI driver include the usb.h `linux/usb.h>` in linux/drivers/usb/host/ehci-hcd.c. Actually I don't understand why the simple helper API can't be called in lower level. Just want to figure out the style, do you mean that only callback is right from lower level to higher level? Anyway I have changed the codes to get device information based on the xfer. When developing the usb host stack, I think the codes may need to be updated to match host stack's design. For now, it should work.

Reply:
> Actually I don't understand why the simple helper API can't be called in lower level.

Because it is always a level violation, and it ends in a nightmare.

>I think the device's related hub information (hub address, port num, first hs hub address, first hs hub port num) and speed are also needed by the controller driver to do the transfer. Will you plan to put all the information to every uhc_transfer?

Yes, but maybe rather a pointer to a device structure similar to `struct urb`. struct usb_device would have to be defined in uhc.h and replace addr in `struct uhc_transfer`.

Reply:
OK, Thanks for the explanation.